src/network/protocol/redis/command/network_protocol_redis_command_mget.c

```diff
@@ -0,0 +1,311 @@
+        goto end;
+    }
+
+    mget_command_context->keys[argument_index] = mget_command_context->key;
+    mget_command_context->key = NULL;
+end:
+    return true;
+}
+
+NETWORK_PROTOCOL_REDIS_COMMAND_FUNCPTR_COMMAND_END(mget) {
+    bool res;
+    bool return_res = false;
+    storage_db_entry_index_t *entry_index = NULL;
+
+    mget_command_context_t *mget_command_context = (mget_command_context_t*)protocol_context->command_context;
+
+    if (mget_command_context->has_error) {
+        char error_send_buffer[256], *error_send_buffer_start, *error_send_buffer_end;
+        size_t error_send_buffer_length;
+        error_send_buffer_length = sizeof(error_send_buffer);
+        error_send_buffer_start = error_send_buffer;
+        error_send_buffer_end = error_send_buffer_start + error_send_buffer_length;
+
+        error_send_buffer_start = protocol_redis_writer_write_simple_error(
+                error_send_buffer_start,
+                error_send_buffer_end - error_send_buffer_start,
+                mget_command_context->error_message,
+                (int)strlen(mget_command_context->error_message));
+
+        if (error_send_buffer_start == NULL) {
+            LOG_E(TAG, "buffer length incorrectly calculated, not enough space!");
+            goto end;
+        }
+
+        return_res = network_send(
+                channel,
+                error_send_buffer,
+                error_send_buffer_start - error_send_buffer) == NETWORK_OP_RESULT_OK;
+
+        goto end;
+    }
+
+    char send_buffer[64], *send_buffer_start, *send_buffer_end;
+    size_t send_buffer_length;
+
+    send_buffer_length = sizeof(send_buffer);
+    send_buffer_start = send_buffer;
+    send_buffer_end = send_buffer_start + send_buffer_length;
+
+    send_buffer_start = protocol_redis_writer_write_array(
+            send_buffer_start,
+            send_buffer_end - send_buffer_start,
+            mget_command_context->keys_count);
+
+    if (send_buffer_start == NULL) {
+        LOG_E(TAG, "buffer length incorrectly calculated, not enough space!");
+        goto end;
+    }
+
+    for(int key_index = 0; key_index < mget_command_context->keys_count; key_index++) {
+        storage_db_entry_index_status_t old_status;
+
+//        LOG_I(TAG, "%s %lu ",mget_command_context->keys[key_index],
+//              strlen(mget_command_context->keys[key_index]));
+
+        entry_index = storage_db_get_entry_index(
+                db,
+                mget_command_context->keys[key_index],
+                strlen(mget_command_context->keys[key_index]));
+
+        if (likely(entry_index)) {
+                // Try to acquire a reader lock until it's successful or the entry index has been marked as deleted
+                storage_db_entry_index_status_increase_readers_counter(
+                        entry_index,
+                        &old_status);
+
+                // TODO: ??
+                if (unlikely(old_status.deleted)) {
```

Comment:
A little review here pls 🙏 idk if i did it right

Reply:
What's happend behind the scenes is that `storage_db_entry_index_status_increase_readers_counter` will internally increase a counter to `prevent` the entry from being deleted (by a SET that replace the key or a DELETE) and ensure that it hasn't been deleted while the counter was being incremented.

Normally `storage_db_get_entry_index` will not return the entry index if deleted but it might get deleted by another thread after it's returned, so there is this extra check.

The `entry_index` returned by `storage_db_get_entry_index` is guaranteed to exist in memory until a fiber switch if it;s not locked by `storage_db_entry_index_status_increase_readers_counter` but it's not guaranteed to do not have been marked deleted by another thread (which means it's going to be deleted very soon if there are readers or within the next fiber switch if there aren't).

So long story short: once the entry index is fetched and the readers counter is increased it's necessary to check if the status fetched from the call is reporting the entry to be deleted.

If it's reported as deleted it's not necessary to decrease the readers counter, this is done automatically internally by `storage_db_entry_index_status_increase_readers_counter`.

If it's deleted it's better to set the entry_index to `NULL`.

The code should always terminate the if and have another if that checks again if `entry_index` is `NULL` so if it set to `NULL` by the previous process the code will flow normally as there would be no mach in the storage db at all.